spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,504 @@
+              + "Row-group merging cannot apply deletes.");
+      return false;
+    }
+
+    // Check if partition spec is being changed - row-group merge cannot repartition data
+    int outputSpecId = group.outputSpecId();
+    boolean hasPartitionSpecChange =
+        group.rewrittenFiles().stream().anyMatch(file -> file.specId() != outputSpecId);
+
+    if (hasPartitionSpecChange) {
+      LOG.debug(
+          "Cannot use row-group merge: partition spec is changing. "
+              + "Row-group merging cannot change data partitioning.");
+      return false;
+    }
+
+    // Check if we're splitting large files - row-group merge is designed for file merging, not
+    // splitting
+    long maxOutputFileSize = group.maxOutputFileSize();
+    boolean isSplittingFiles =
+        group.rewrittenFiles().stream()
+            .anyMatch(file -> file.fileSizeInBytes() > maxOutputFileSize);
+
+    if (isSplittingFiles) {
+      LOG.debug(
+          "Cannot use row-group merge: compaction is splitting large files. "
+              + "Row-group merging is optimized for merging files, not splitting them.");
+      return false;
+    }
+
+    // Validate schema compatibility and other Parquet-specific requirements
+    try {
+      List<InputFile> inputFiles =
+          group.rewrittenFiles().stream()
+              .map(f -> table().io().newInputFile(f.path().toString()))
+              .collect(Collectors.toList());
+
+      // Validate files can be merged
+      boolean canMerge = ParquetFileMerger.canMerge(inputFiles);
+
+      if (!canMerge) {
+        LOG.info(
+            "Cannot use row-group merge for {} files. Falling back to standard rewrite. "
+                + "Reason: Parquet validation failed",
+            group.rewrittenFiles().size());
+      }
+
+      return canMerge;
+    } catch (Exception e) {
+      LOG.info("Cannot use row-group merge: validation failed", e);
+      return false;
+    }
+  }
+
+  /**
+   * Merges Parquet files in the group, respecting the expected output file count determined by the
+   * planner. Files are distributed evenly across the expected number of output files.
+   */
+  private void mergeParquetFilesDistributed(String groupId, RewriteFileGroup group) {
+    PartitionSpec spec = table().specs().get(group.outputSpecId());
+    StructLike partition = group.info().partition();
+    int expectedOutputFiles = group.expectedOutputFiles();
+
+    LOG.info(
+        "Merging {} Parquet files into {} expected output files (group: {})",
+        group.rewrittenFiles().size(),
+        expectedOutputFiles,
+        groupId);
+
+    // Check if table supports row lineage
+    boolean preserveRowLineage = TableUtil.supportsRowLineage(table());
+
+    // Distribute files evenly across expected output files (planner already determined the count)
+    List<List<DataFile>> fileBatches =
+        distributeFilesEvenly(group.rewrittenFiles(), expectedOutputFiles);
```

Comment:
What if we force the user to provide parameters to the original planner in a way, that it creates groups which are only result in a single expectedOutputFile. We can fail here if the `expectedOutputFiles` is bigger than 1. This way we can get rid a bunch of code, and simplify things quite a bit.

Reply:
Thanks for the suggestion! I understand it is important to simplify the code, but I have concerns about forcing expectedOutputFiles = 1 because that will create tight coupling between the planner and runner implementation. Also, with the limit of max concurrent file group rewrites (MAX_CONCURRENT_FILE_GROUP_REWRITES_DEFAULT), not sure it would impact the Parallelism.

Reply:
The user should set `MAX_FILE_GROUP_SIZE_BYTES` to `TARGET_FILE_SIZE_BYTES`.
In this case all of the groups created will result exactly one output file.

We could just check if there is only a single expectedOutputFile and log and fall back to the normal merge if there are more.

The RewriteDataFilesSparkAction could check the config and fail/warn if it is not correct

Reply:
If expectedOutputFiles = 1, the planner will end up producing more groups. Could you cast a light on how these groups are parallelized during execution?

Reply:
- Added validation in canMerge() to require expectedOutputFiles == 1
  - Groups with multiple expected files gracefully fall back to standard bin-pack rewrite
  - Removed distributeFilesEvenly() bin-packing method (45 lines)
  - Simplified batching logic: removed loops for creating tasks and processing results